string.c

```diff
@@ -2732,20 +2751,14 @@ str_subseq(VALUE str, long beg, long len)
 {
     VALUE str2;
 
-    const long rstring_embed_capa_max = ((sizeof(struct RString) - offsetof(struct RString, as.embed.ary)) / sizeof(char)) - 1;
-
-    if (!SHARABLE_SUBSTRING_P(beg, len, RSTRING_LEN(str)) ||
-            len <= rstring_embed_capa_max) {
+    if (!SHARABLE_SUBSTRING_P(beg, len, RSTRING_LEN(str))) {
         str2 = rb_str_new(RSTRING_PTR(str) + beg, len);
         RB_GC_GUARD(str);
     }
     else {
-        str2 = str_new_shared(rb_cString, str);
+        str2 = str_new_shared_suffix(rb_cString, str, beg);
```

Comment:
I pushed `len` down to `str_new_shared_partial` (renamed from `str_new_shared_suffix`), too.  Now `str_replace_shared_partial_without_enc` will make shared strings in the middle, too, if the macro `SHARABLE_SUBSTRING_P` is 1.  In the end of `str_replace_shared_partial_without_enc`, the statement

```c
    // Excerpt from str_replace_shared_partial_without_enc
    STR_SET_LEN(str2, partial_len);
```

will set the string length to the length of the sub-string so that it will not "go all the way to the end".